ankaios_sdk/ankaios.py

```diff
@@ -0,0 +1,662 @@
+                             + "is not started.")
+
+        with self._responses_lock:
+            if request_id in self._responses:
+                return self._responses.pop(request_id).get_response()
+            self._responses[request_id] = ResponseEvent()
+
+        return self._responses[request_id].wait_for_response(timeout)
+
+    def _write_to_pipe(self, request: Request) -> None:
+        """
+        Writes the request into the control interface output fifo.
+
+        Args:
+            request (Request): The request object to be written.
+        """
+        with open(f"{self.ANKAIOS_CONTROL_INTERFACE_BASE_PATH}/output",
+                  "ab") as f:
+            request_to_ankaios = _control_api.ToAnkaios(
+                request=request._to_proto()
+            )
+            # Send the byte length of the proto msg
+            f.write(_VarintBytes(request_to_ankaios.ByteSize()))
+            # Send the proto msg itself
+            f.write(request_to_ankaios.SerializeToString())
+            f.flush()
+
+    def _send_request(self, request: Request,
+                      timeout: float = DEFAULT_TIMEOUT) -> Response:
+        """
+        Send a request and wait for the response.
+
+        Args:
+            request (Request): The request object to be sent.
+            timeout (float): The maximum time to wait for the response,
+                in seconds.
+
+        Returns:
+            Response: The response object.
+        """
+        if not self._connected:
+            raise ValueError("Cannot request if not connected.")
+        self._write_to_pipe(request)
+
+        try:
+            response = self._get_response_by_id(request.get_id(), timeout)
+        except TimeoutError as e:
+            raise e
+        return response
+
+    def set_logger_level(self, level: AnkaiosLogLevel) -> None:
+        """
+        Set the log level of the logger.
+
+        Args:
+            level (AnkaiosLogLevel): The log level to be set.
+        """
+        self.logger.setLevel(level.value)
+
+    def connect(self) -> None:
+        """
+        Connect to the control interface by starting to read
+        from the input fifo.
+
+        Raises:
+            ValueError: If already connected.
+        """
+        if self._connected:
+            raise ValueError("Already connected.")
+        self._connected = True
+        self._read_thread = threading.Thread(
+            target=self._read_from_control_interface
+        )
+        self._read_thread.start()
+
+    def disconnect(self) -> None:
+        """
+        Disconnect from the control interface by stopping to read
+        from the input fifo.
+
+        Raises:
+            ValueError: If already disconnected.
+        """
+        if not self._connected:
+            raise ValueError("Already disconnected.")
+        self._connected = False
+        self._read_thread.join()
+
+    def apply_manifest(self, manifest: Manifest) -> None:
```

Comment:
All methods making changes to Ankaios shall give an indication what actually happened.
Especially the success message is important, as it returns the ids of the changed workloads. We should also consider how we formulate the error as the Control Interface now also supports closing the connection from the server side.

To allow the user to wait for a certain workload state to be reached, we can also provide a method that accepts a list of workload instance names and an execution state (and a timeout) and returns only when the state is reached. 
When I think about it now, to support more complex workflows, we should also provide a `get workload state for instance name`.

Reply:
I will change the method to return a bool indicating if it worked or not. The logs are already present and document what happened, in each case.

As for the WorkloadInstanceName proposal: There is no use-case for that apart from accessing it in the Workload State that you get when requesting the state of the system. You are not able to use it otherwise.
The rationale of this is that you cannot create a WorkloadinstanceName because it requires the workload id. I have designed all the method to accept workload name and/or agent name and return a list with the found workloads (that may have the same name / same agent name or both, but not the same id). I am not sure if this is the right / easiest approach. Every suggestion to how this should be handled better is appreciated.

Reply:
A list of WorkloadInstanceNames is returned by each UpdateWorkload triggered on the APIs of Ankaios as part of the success message:
```proto
/**
* A message from the server containing the ids of the workloads that have been started and stopped in response to a previously sent UpdateStateRequest.
*/
message UpdateStateSuccess {
    repeated string addedWorkloads = 1; /// Workload istance names of workloads which will be started
    repeated string deletedWorkloads = 2; /// Workload instance names of workloads which will be stopped
}
```

The idea is that you trigger the update and Ankaios tells you what will be changed. Using the Ids you can monitor the progress of the update.

Reply:
Currently, the content of that response is checked if it's ok or not. It's printed in the logger either way. But it is not returned from the method. The user only gets a boolean (as mentioned before) if it's ok or not. I believe this should be redone, with the result better handled.

Reply:
Changed so that the user will get lists with the added and deleted workloads as `WorkloadInstanceName` objects and added functionality to better incorporate the uses of the objects.